mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp

```diff
@@ -177,6 +178,81 @@ void affixGridwiseGemmAttributes(Operation *convOp, Operation *gop,
   }
 }
 
+/// 0-initialize the output for a backward weight convolution which uses
+/// atomic adds.
+/// For f32 type, the output is the filter tensor.
+/// for f16 type, the output is the workspace.
+LogicalResult zeroInit(Conv2DBwdWeightOp op, PatternRewriter &b) {
+  auto loc = op.getLoc();
+  auto filterDataType =
+      op.filter().getType().cast<MemRefType>().getElementType();
+  Value output;
+  if (filterDataType == b.getF32Type()) {
+    output = op.filter();
+  } else if (filterDataType == b.getF16Type()) {
+    assert(op.workspace() && "Op has no workspace");
+    output = op.workspace();
+  } else {
+    llvm_unreachable("Incorrect memref type supplied");
+  }
+  auto outputDataType = output.getType().cast<MemRefType>().getElementType();
+  auto zeroOp = createZeroConstantOp(b, loc, outputDataType);
+  auto collapsedOutput = createCollapseShapeOp(b, loc, output);
+  ArrayRef<int64_t> collapsedOutputShape =
+      collapsedOutput.getType().cast<MemRefType>().getShape();
+  llvm::SmallVector<AffineForOp, 1> affineLoops;
+  for (auto bound : collapsedOutputShape) {
+    auto loop = b.create<AffineForOp>(loc, 0, bound);
+    affineLoops.push_back(loop);
+    auto body = loop.getBody();
+    b.setInsertionPointToStart(body);
+  }
+  llvm::SmallVector<Value, 1> indices;
+  extractForInductionVars(affineLoops, &indices);
+  b.create<AffineStoreOp>(loc, zeroOp, collapsedOutput, indices);
+
+  op.erase();
+  return success();
+}
+
+/// Element-wise conversion from the workspace to the output (filter tensor)
+/// for a backward weight convolution which uses atomic adds.
+LogicalResult elementwiseConversion(Conv2DBwdWeightOp op, PatternRewriter &b) {
+  auto loc = op.getLoc();
+  assert(op.workspace() && "Op has no workspace");
+  auto filter = op.filter();
+  auto workspace = op.workspace();
+  auto filterDataType = filter.getType().cast<MemRefType>().getElementType();
+  auto workspaceDataType =
+      workspace.getType().cast<MemRefType>().getElementType();
+  auto collapsedFilter = createCollapseShapeOp(b, loc, filter);
+  auto collapsedWorkspace = createCollapseShapeOp(b, loc, workspace);
```

Comment:
Why do we need this?

Reply:
We are certain we do an element-wise conversion, so we can use `memref.collapse_shape` to make it a 1D memref to simplify further code emissions.